OpenPGP signature creation with external signing code

btoews opened the pull request:
We're wanting to make OpenPGP signatures using a HSM to do the actual RSA signing via PKCS#11. This isn't feasible (possible?) with the current APIs.

This PR adds a new `Signer` interface, allowing application code to decide how the actual signature is created:

``` go
type Signer interface {
    Sign(sig *Signature, msg []byte) error
    KeyId() uint64
    PublicKeyAlgo() PublicKeyAlgorithm
}
```

It also adds `openpgp.SignWithSigner` and `packet.Signature#SignWithSigner` functions for making detached signatures using the new interface. The code for these is mostly borrowed from `openpgp.detachSign` and `packet.Signature#Sign`, except the actual signing is delegated to the `Signer`.

I tried to make this general enough that it would work for other use cases. I'm curious what others think.

/cc @brianmario

mastahyeti committed:
add openpgpg.SignWithSigner

btoews commented:
I'm refactoring this a bit. Sorry for the false alarm.

mastahyeti committed:
make Signer act like hash.Hash

mastahyeti committed:
allow signing keys and uids too

brianmario commented:
Nice! Yeah I like the use of the `hash.Hash` interface.

btoews commented:
I finished the refactoring I was doing, so this should be ready for :eyes:. I made the `Signer` interface include the `hash.Hash` interface, which simplifies things a bit.

btoews commented:
Bump. Anyone have thoughts on whether this API addition makes sense? We'd love to not have to maintain our own fork of this code...

maxtaco commented:
Thanks @mastahyeti this looks pretty good.  Can you add a little bit more test coverage for some of the error cases you introduced? Thanks!

mastahyeti committed:
keySignatureHash doesn't return an error any more

mastahyeti committed:
re-abstract the error handling for unavailable hashes

mastahyeti committed:
do the same for userIdSignatureHash

mastahyeti committed:
revert some of the naming changes

mastahyeti committed:
revert some ws changes

btoews commented:
> Can you add a little bit more test coverage for some of the error cases you introduced?

Most of the error cases in the diff were added because I extracted the error checking out of `keySignatureHash` and `userIdSignatureHash` and moved it to the calling functions. I refactored this a bit to keep the error checking in those functions and to keep their signatures the same.

Keeping the function signatures the  same should make this fork easier to maintain when you pull in updates from the upstream repo. The other API change in this PR is renaming `fromBig` to `FromBig`. I'm wondering if instead, I should add a new exported function `MpiFromBig` that simply calls `fromBig` to avoid that API change too (like [this](https://github.com/keybase/go-crypto/commit/84e498402210bbe3b7dc6232fd928c0f5cedf024)). Thoughts?

As for tests, I'm a bit uncertain of what needs to be added. Are there any specific cases you'd like to see tested?

maxtaco commented on openpgp/packet/signature.go:
```diff
@@ -539,7 +547,9 @@ func (sig *Signature) signPrepareHash(h hash.Hash) (digest []byte, err error) {
-	if priv.PrivateKey == nil {
+	signer, hashIsSigner := h.(Signer)
+
+	if !hashIsSigner && priv.PrivateKey == nil {
```
I was wondering if it's possible to make a test case for this error condition....

btoews replied:
I added a test in b5bb787789c47b9d53f45610eeb3cdcc8195afb6, checking that passing a non-Signer `h` and a `nil` `priv` will trigger this error.

maxtaco commented on openpgp/packet/signature.go:
```diff
@@ -550,6 +560,11 @@ func (sig *Signature) Sign(h hash.Hash, priv *PrivateKey, config *Config) (err e
 	}
 
+	if hashIsSigner {
+		err = signer.Sign(sig)
```
...and this one here, though, not having followed it all the way through, I understand what you're saying that these might not be new error cases...

btoews replied:
I added a test in 21bb752d661b05f3c87a1a302b920b902b517ce0, checking that errors returned from `Signer.Sign` will be bubbled up through `Signature.Sign` and `write.SignWithSigner`.

mastahyeti committed:
test nil `priv` argument to Sign method

mastahyeti committed:
test that errors from Signer.Sign are bubbled up

maxtaco commented:
Thanks!

maxtaco commented on openpgp/packet/signature.go:
```diff
@@ -605,6 +630,15 @@ func (sig *Signature) SignKey(pub *PublicKey, priv *PrivateKey, config *Config)
 	return sig.Sign(h, priv, config)
 }
 
+// SignKey computes a signature from priv, asserting that pub is a subkey. On
```
fix comment here to match function name

maxtaco commented on openpgp/write.go:
```diff
@@ -59,6 +59,35 @@ func armoredDetachSign(w io.Writer, signer *Entity, message io.Reader, sigType p
 	return out.Close()
 }
 
+func SignWithSigner(s packet.Signer, w io.Writer, message io.Reader, sigType packet.SignatureType, config *packet.Config) (err error) {
```
Can you provide a comment on this function?

maxtaco commented:
Good to merge after some comment fixes. Thanks!

mastahyeti committed:
update docs

btoews commented:
I updated the comments in da273caefaf61a6b4c2c695405b7b82c4c0ec88b.

btoews commented:
:metal: Thanks @maxtaco

maxtaco commented:
Thank you!
